substrate/frame/balances/src/lib.rs

```diff
@@ -232,6 +232,17 @@ pub mod pallet {
 
 			type WeightInfo = ();
 		}
+
+		/// Default configurations of this pallet in a solochain environment.
+		pub struct SolochainDefaultConfig;
+
+		#[derive_impl(frame_system::config_preludes::SolochainDefaultConfig, no_aggregated_types)]
+		impl frame_system::DefaultConfig for SolochainDefaultConfig {}
+
+		/// It currently uses the same configuration as [`TestDefaultConfig`].
+		#[derive_impl(TestDefaultConfig, no_aggregated_types)]
+		#[frame_support::register_default_impl(SolochainDefaultConfig)]
+		impl DefaultConfig for SolochainDefaultConfig {}
```

Comment:
Why do pallets need to be concerned with the default config of the solochain?

Shouldn't config related logic for the solochain be in the solochain crate?

Where does this stop? Will we start adding default configs to every pallet for all our 4,5,6,7+ runtime variations?

Reply:
> Shouldn't config related logic for the solochain be in the solochain crate?

Which crate are you referring here? Is the expectation then for the runtime dev to import that crate as well? How does that reduce the friction?

> Where does this stop? Will we start adding default configs to every pallet for all our 4,5,6,7+ runtime variations?

I would imagine that we provide a few (Test, Solo, Para and Relay) and let the developers build out the rest as needed for custom use cases.

Reply:
My initial reaction to this is I really don't like it. 

1. I don't think pallet developers should be concerned about needing to write a default config for every runtime variation the pallet could be used in, the entire point of pallets exposing a config is so the customisation happens in the **runtime**?

2. This also encourages lazy, or at best 'blind' runtime development. Critical configuration of your runtime is decided for you in a crates.io package you'll probably never look at. Even if you do check the downstream package, what if a new default config item is added to the downstream package in an upgrade that doesn't suit your runtime? It will be added without you even knowing. 

I'm all for default configs in the context of testing and mocking, but I think for production systems with value on the line it is much better for runtime developers to explicitly specify their configs.

Reply:
I think this is a longer discussion outside the scope of this PR. Could you open a separate issue to take that further?

Reply:
OK. Before merging this PR I don't think the production chain config should extend the test config though.

Changing the test config seems like it would be an innocent change, I wouldn't expect that to impact production chains.

Reply:
I agree inheriting from the testconfig could be very bad.
Sane defaults (as restrictive as possible) for not having to always type out unreasonably large config impls is what we need. We don't want only mock runtimes to benefit from this.
If we agree, I would only have a `TestDefaultConfig` and a `RestrictiveDefaultConfig` or something like that.

Reply:
This was done based on this comment: https://github.com/paritytech/polkadot-sdk/pull/3946#discussion_r1548062675

Since this is not the core focus of this PR, I will revert it so that this can move ahead. We can continue this discussion on the original issue.

Reply:
This is done. Thanks.